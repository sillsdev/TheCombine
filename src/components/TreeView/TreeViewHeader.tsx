import React from "react";
import {
  Button,
  Typography,
  TextField,
  Grid,
  GridList,
  GridListTile,
  Card,
} from "@material-ui/core";
import ChevronLeft from "@material-ui/icons/ChevronLeft";
import ChevronRight from "@material-ui/icons/ChevronRight";

import SemanticDomainWithSubdomains from "./SemanticDomain";

interface TreeHeaderProps {
  currentDomain: SemanticDomainWithSubdomains;
  animate: (domain: SemanticDomainWithSubdomains) => Promise<void>;
}

interface TreeHeaderState {
  input: string;
}

export default class TreeViewHeader extends React.Component<
  TreeHeaderProps,
  TreeHeaderState
> {
  animating: boolean;

  constructor(props: TreeHeaderProps) {
    super(props);
    this.state = {
      input: props.currentDomain.id,
    };

    this.animating = false;

    this.searchAndSelectDomain = this.searchAndSelectDomain.bind(this);
    this.navigateDomainArrowKeys = this.navigateDomainArrowKeys.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.updateDomain = this.updateDomain.bind(this);
  }

  componentDidMount() {
    window.addEventListener("keydown", this.navigateDomainArrowKeys);
  }

  componentWillUnmount() {
    window.removeEventListener("keydown", this.navigateDomainArrowKeys);
  }

  // Change the input on typing
  handleChange(event: React.ChangeEvent<HTMLTextAreaElement>) {
    this.setState({ input: event.target.value });
  }

  // Dispatch the search for a specified domain, and switches to it if it exists
  searchAndSelectDomain(event: React.KeyboardEvent) {
    event.bubbles = false;
    event.preventDefault();

    if (event.key === "Enter") {
      // Find parent domain
      let parent: SemanticDomainWithSubdomains | undefined = this.props
        .currentDomain;
      while (parent.parentDomain !== undefined) parent = parent.parentDomain;

      // Search for domain
      if (!isNaN(parseInt(this.state.input)))
        for (let i: number = 0; i < 4 && parent; i++) {
          parent = this.searchDomainByNumber(
            parent,
            this.state.input.slice(0, i * 2 + 1)
          );
          if (parent && parent.id === this.state.input) {
            this.props.animate(parent);
            this.setState({ input: "" });
            (event.target as any).value = "";
          } else if (!parent) break;
        }
      else {
        parent = this.searchDomainByName(parent, this.state.input);
        if (parent) {
          this.props.animate(parent);
          this.setState({ input: "" });
          (event.target as any).value = "";
        }
      }
    }
  }

  // Navigate tree via arrow keys
  navigateDomainArrowKeys(event: KeyboardEvent) {
    if (event.key === "ArrowLeft") {
      let domain:
        | SemanticDomainWithSubdomains
        | undefined = this.getBrotherDomain(-1);
      if (domain && domain.id !== this.props.currentDomain.id)
        this.props.animate(domain);
    } else if (event.key === "ArrowRight") {
      let domain:
        | SemanticDomainWithSubdomains
        | undefined = this.getBrotherDomain(1);
      if (domain && domain.id !== this.props.currentDomain.id)
        this.props.animate(domain);
    } else if (event.key === "ArrowDown") {
      if (this.props.currentDomain.parentDomain)
        this.props.animate(this.props.currentDomain.parentDomain);
    }
  }

  // Search for a semantic domain by number
  searchDomainByNumber(
    parent: SemanticDomainWithSubdomains,
    number: string
  ): SemanticDomainWithSubdomains | undefined {
    for (let domain of parent.subdomains)
      if (domain.id === number) return domain;

    if (parent.id === number) return parent;
    else return undefined;
  }

  // Searches for a semantic domain by name
  searchDomainByName(
    domain: SemanticDomainWithSubdomains,
    target: string
  ): SemanticDomainWithSubdomains | undefined {
    let check = (checkAgainst: SemanticDomainWithSubdomains | undefined) =>
      checkAgainst && target.toLowerCase() === checkAgainst.name.toLowerCase();
    if (check(domain)) return domain;

    // If there are subdomains
    if (domain.subdomains.length > 0) {
      let tempDomain: SemanticDomainWithSubdomains | undefined;
      for (let sub of domain.subdomains) {
        tempDomain = this.searchDomainByName(sub, target);
        if (check(tempDomain)) return tempDomain;
      }
    }
  }

  // Switches currentDomain to the domain navigationAmount off from this domain, assuming that domain exists
  navigateDomain(navigationAmount: number) {
    if (this.props.currentDomain.parentDomain) {
      let brotherDomain:
        | SemanticDomainWithSubdomains
        | undefined = this.getBrotherDomain(navigationAmount);
      if (brotherDomain) this.props.animate(brotherDomain);
    }
  }

  // Gets the domain 'navigationAmount' away from the currentDomain (negative to the left, positive to the right)
  getBrotherDomain(
    navigationAmount: number
  ): SemanticDomainWithSubdomains | undefined {
    if (this.props.currentDomain.parentDomain) {
      let brotherDomains: SemanticDomainWithSubdomains[] = this.props
        .currentDomain.parentDomain.subdomains;
      let index: number = brotherDomains.findIndex(
        (domain) => this.props.currentDomain.id === domain.id
      );

      index += navigationAmount;
      if (0 <= index && index < brotherDomains.length)
        return brotherDomains[index];
    }

    // No brother domain navigationAmount over from currentDomain
    return undefined;
  }

  // Switches current semantic domain + updates search bar
  updateDomain() {
    this.setState({ input: this.props.currentDomain.id });
  }

  // Creates the L/R button + select button + search bar
  render() {
    let domainL:
      | SemanticDomainWithSubdomains
      | undefined = this.getBrotherDomain(-1);
    let domainR:
      | SemanticDomainWithSubdomains
      | undefined = this.getBrotherDomain(1);

    return (
      <GridList cols={5} spacing={2} cellHeight={"auto"}>
        <GridListTile cols={1}>
          {domainL ? (
            <Grid container justify="center">
              <Button
                variant={"outlined"}
                onClick={() => this.navigateDomain(-1)}
                style={{ float: "right", marginTop: "50%" }}
              >
                <ChevronLeft />
                <Typography variant="body2">
                  {domainL.id.padStart(8, " ")}
                </Typography>
              </Button>
            </Grid>
          ) : null}
        </GridListTile>
        <GridListTile cols={3}>
          <Card>
            <Button
              fullWidth
              size="large"
              color="primary"
              variant="contained"
              disabled={!this.props.currentDomain.parentDomain}
              onClick={() => this.props.animate(this.props.currentDomain)}
            >
              <div style={{ textTransform: "capitalize" }}>
                <Typography variant="overline">
                  {this.props.currentDomain.id}
                </Typography>
                <Typography variant="h6">
                  {this.props.currentDomain.name}
                </Typography>
              </div>
            </Button>
            <TextField
              fullWidth
              id="name"
              label="Find a domain"
              onKeyUp={this.searchAndSelectDomain}
              onChange={this.handleChange}
              margin="normal"
              autoComplete="off"
            />
          </Card>
        </GridListTile>
        <GridListTile cols={1}>
          {domainR ? (
            <Grid container justify="center">
              <Button
                variant={"outlined"}
                onClick={() => this.navigateDomain(1)}
                style={{ marginTop: "50%" }}
              >
                <Typography variant="body2">
                  {domainR.id.padEnd(8, " ")}
                </Typography>
                <ChevronRight />
              </Button>
            </Grid>
          ) : null}
        </GridListTile>
      </GridList>
    );
  }
}
