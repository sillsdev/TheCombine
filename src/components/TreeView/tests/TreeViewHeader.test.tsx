import React from "react";
import renderer, { ReactTestRenderer } from "react-test-renderer";
import SemanticDomainWithSubdomains from "../SemanticDomain";
import TreeViewHeader from "../TreeViewHeader";
import MockTree from "./MockSemanticTree";

// Variable event
var event = {
  bubbles: false,
  key: "Enter",
  preventDefault: jest.fn(),
  target: {
    value: "",
  },
};

// Handles
var treeMaster: ReactTestRenderer;
var treeHandle: TreeViewHeader;
const MOCK_ANIMATE = jest.fn();

// Mock out window
const ADD_LISTENER = window.addEventListener.prototype;
const REMOVE_LISTENER = window.removeEventListener.prototype;
window = {
  ...window,
  addEventListener: jest.fn(),
  removeEventListener: jest.fn(),
};

beforeEach(() => {
  setTree(MockTree.subdomains[1]);
  MOCK_ANIMATE.mockClear();
});

afterAll(() => {
  window = {
    ...window,
    addEventListener: ADD_LISTENER,
    removeEventListener: REMOVE_LISTENER,
  };
});

describe("Tests TreeViewHeader", () => {
  it("Renders correctly", () => {
    // Default snapshot test
    snapTest("default view");
  });

  // onKeyDown
  it("Search & select domain switches semantic domain if given number found", () => {
    treeHandle.setState({ input: MockTree.id });
    event.target.value = "not empty";
    treeHandle.searchAndSelectDomain((event as any) as React.KeyboardEvent);

    expect(MOCK_ANIMATE).toHaveBeenCalledWith(MockTree);
    expect(treeHandle.state.input).toEqual("");
    expect(event.target.value).toEqual("");
  });

  it("Search & select domain does not switch semantic domain if given number not found", () => {
    const TEST: string = "10";
    treeHandle.setState({ input: TEST });
    event.target.value = TEST;
    treeHandle.searchAndSelectDomain((event as any) as React.KeyboardEvent);

    expect(MOCK_ANIMATE).toHaveBeenCalledTimes(0);
    expect(treeHandle.state.input).toEqual(TEST);
    expect(event.target.value).toEqual(TEST);
  });

  it("Search & select domain switches semantic domain if given name found", () => {
    treeHandle.setState({ input: MockTree.subdomains[2].name });
    event.target.value = "not empty";
    treeHandle.searchAndSelectDomain((event as any) as React.KeyboardEvent);

    expect(MOCK_ANIMATE).toHaveBeenCalledWith(MockTree.subdomains[2]);
    expect(treeHandle.state.input).toEqual("");
    expect(event.target.value).toEqual("");
  });

  it("Search & select domain does not switch semantic domain if given name not found", () => {
    const TEST: string = "itsatrap";
    treeHandle.setState({ input: TEST });
    event.target.value = TEST;
    treeHandle.searchAndSelectDomain((event as any) as React.KeyboardEvent);

    expect(MOCK_ANIMATE).toHaveBeenCalledTimes(0);
    expect(treeHandle.state.input).toEqual(TEST);
    expect(event.target.value).toEqual(TEST);
  });

  // getBrotherDomain
  it("provides the proper brother domains", () => {
    // Standard navigation
    expect(treeHandle.getBrotherDomain(-1)).toEqual(MockTree.subdomains[0]);
    expect(treeHandle.getBrotherDomain(1)).toEqual(MockTree.subdomains[2]);

    // Check with indices out-of-bounds
    expect(treeHandle.getBrotherDomain(-2)).toEqual(undefined);
    expect(treeHandle.getBrotherDomain(2)).toEqual(undefined);

    // Check that a domain w/ no parentDomain has no brotherDomains
    setTree(MockTree);
    expect(treeHandle.getBrotherDomain(-1)).toEqual(undefined);
  });

  // navigateKeys
  it("navigateKeys w/ Arrow Down switches to parent domain", () => {
    event.key = "ArrowDown";
    treeHandle.navigateDomainArrowKeys((event as any) as KeyboardEvent);
    expect(MOCK_ANIMATE).toHaveBeenCalledWith(MockTree);
  });

  it("navigateKeys w/ Arrow Left switches to left brother domain", () => {
    event.key = "ArrowLeft";
    treeHandle.navigateDomainArrowKeys((event as any) as KeyboardEvent);
    expect(MOCK_ANIMATE).toHaveBeenCalledWith(MockTree.subdomains[0]);
  });

  it("navigateKeys w/ Arrow Right switches to right brother domain", () => {
    event.key = "ArrowRight";
    treeHandle.navigateDomainArrowKeys((event as any) as KeyboardEvent);
    expect(MOCK_ANIMATE).toHaveBeenCalledWith(MockTree.subdomains[2]);
  });
});

// Creates the tree
function setTree(domain: SemanticDomainWithSubdomains) {
  renderer.act(() => {
    treeMaster = renderer.create(
      <TreeViewHeader currentDomain={domain} animate={MOCK_ANIMATE} />
    );
  });
  treeHandle = treeMaster.root.findByType(TreeViewHeader).instance;
}

// Perform a snapshot test
function snapTest(name: string) {
  expect(treeMaster.toJSON()).toMatchSnapshot();
}
