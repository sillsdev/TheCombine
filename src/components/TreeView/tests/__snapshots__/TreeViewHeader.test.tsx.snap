// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests TreeViewHeader Renders correctly 1`] = `
<div
  className="MuiGrid-root MuiGrid-container"
>
  <div
    className="MuiGrid-root MuiGrid-item"
    style={
      Object {
        "minWidth": "5vw",
      }
    }
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-outlined"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <h6
          className="MuiTypography-root MuiTypography-h6"
        >
          &lt;     1.0
        </h6>
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item"
    style={
      Object {
        "minWidth": "10vw",
      }
    }
  >
    <div
      className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
      onKeyUp={[Function]}
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
        data-shrink={true}
        htmlFor="name"
      >
        Semantic Domain
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          className="MuiInputBase-input MuiInput-input"
          disabled={false}
          id="name"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value="1.1"
        />
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item"
    style={
      Object {
        "minWidth": "5vw",
      }
    }
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-outlined"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <h6
          className="MuiTypography-root MuiTypography-h6"
        >
          1.2     &gt;
        </h6>
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
